File: Lista/Java/Q14.java
class Configuracao {
    private static Configuracao instancia;

    private String ambiente;
    private String bancoDeDados;

    private Configuracao() {
        this.ambiente = "Desenvolvimento";
        this.bancoDeDados = "MySQL";
    }

    public static Configuracao getInstancia() {
        if (instancia == null) {
            instancia = new Configuracao();
        }
        return instancia;
    }

    public String getAmbiente() {
        return ambiente;
    }

    public void setAmbiente(String ambiente) {
        this.ambiente = ambiente;
    }

    public String getBancoDeDados() {
        return bancoDeDados;
    }

    public void setBancoDeDados(String bancoDeDados) {
        this.bancoDeDados = bancoDeDados;
    }
}

public class Main {
    public static void main(String[] args) {
        Configuracao config1 = Configuracao.getInstancia();
        System.out.println("Ambiente: " + config1.getAmbiente());
        System.out.println("Banco de Dados: " + config1.getBancoDeDados());

        config1.setAmbiente("Produção");
        config1.setBancoDeDados("PostgreSQL");

        Configuracao config2 = Configuracao.getInstancia();
        System.out.println("\nApós modificações:");
        System.out.println("Ambiente: " + config2.getAmbiente());
        System.out.println("Banco de Dados: " + config2.getBancoDeDados());

        System.out.println("\nAs instâncias são iguais? " + (config1 == config2));
    }
}

File: Lista/Go/Q14.go
package main

import (
    "fmt"
    "sync"
)

type Configuracao struct {
    Ambiente string
    Versao   string
}

var instance *Configuracao

var mu sync.Mutex

func GetConfiguracao() *Configuracao {
    mu.Lock()
    defer mu.Unlock()

    if instance == nil {
        instance = &Configuracao{
            Ambiente: "Desenvolvimento",
            Versao:   "1.0.0",
        }
    }
    return instance
}

func main() {
    config1 := GetConfiguracao()
    fmt.Printf("Configuração 1: Ambiente=%s, Versão=%s\n", config1.Ambiente, config1.Versao)

    config2 := GetConfiguracao()
    fmt.Printf("Configuração 2: Ambiente=%s, Versão=%s\n", config2.Ambiente, config2.Versao)

    if config1 == config2 {
        fmt.Println("As duas instâncias são as mesmas.")
    }
}
